### PR TITLE
  Add slt in template plugin/tensor iterator

### PR DESCRIPTION
**Pull request by wilson-seok**
### Details:
 - Add op reference test of TensorIterator
 - Add visitor api test of TensorIterator

### Tickets:
 - 63999

### CODE REVIEW

File: docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp [hunk 1]
```diff
@@ -0,0 +1,2306 @@
+// Copyright (C) 2021 Intel Corporation
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp]: Please update copyrights

File: docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp [hunk 2]
```diff
@@ -0,0 +1,2306 @@
...
+        const element::Type_t& iType,
+        const std::vector<T>& XValues, const std::vector<T>& H_tValues,  const std::vector<T>& C_tValues, const std::vector<int64_t>& S_tValues,
+        const std::vector<T>& WValues, const std::vector<T>& RValues, const std::vector<T>& BValues,
+        const std::vector<T>& YValues, const std::vector<T>& HoValues, const std::vector<T>& CoValues,
+        const std::string& testcaseName = "") :
+        batchSize(batchSize), inputSize(inputSize), hiddenSize(hiddenSize), seqLength(seqLength),
+        clip(clip), body_type(body_type), direction(direction), iType(iType), oType(iType),
+        testcaseName(testcaseName) {
+            switch (body_type) {
+                case ngraph::helpers::TensorIteratorBody::LSTM: {
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp]: Could you use enums and classes from ov namespace?

File: src/core/tests/CMakeLists.txt [hunk 3]
```diff
@@ -402,6 +402,7 @@ set(SRC
     visitors/op/swish.cpp
     visitors/op/tan.cpp
     visitors/op/tanh.cpp
+    visitors/op/ti.cpp
```
Comment (ilyachur) [inline src/core/tests/CMakeLists.txt:405]: Could you rename it to tensor_iterator?

Comment (wilson-seok) [reply src/core/tests/CMakeLists.txt]: Fixed it! Thanks!

File: docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp (hunk 2 above)
Comment (wilson-seok) [reply docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp]: Fixed it! I used local enum for only this file. Thanks!

File: docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp (hunk 1 above)
Comment (wilson-seok) [reply docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp]: Updated it! Thanks!

**Comment by wilson-seok** [comment]
@ilyachur @iefode I met a confliction of tensor_iterator.cpp with https://github.com/openvinotoolkit/openvino/pull/9752. The PR created new op reference test in template plugin(docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp). Then do I need to merge so both test cases exist in the file? Or close this PR without merge?

**Comment by ilyachur** [comment]
> @ilyachur @iefode I met a confliction of tensor_iterator.cpp with #9752. The PR created new op reference test in template plugin(docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp). Then do I need to merge so both test cases exist in the file? Or close this PR without merge?

Please check coverage of tests and if we need to have more test cases, please extend tests.

**Comment by wilson-seok** [comment]
> > @ilyachur @iefode I met a confliction of tensor_iterator.cpp with #9752. The PR created new op reference test in template plugin(docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp). Then do I need to merge so both test cases exist in the file? Or close this PR without merge?
> 
> Please check coverage of tests and if we need to have more test cases, please extend tests.

@ilyachur My approach is similar with single layer test of functional shared_test_classes. It also uses recurrence ops. The new one is similar with single layer test of cpu plugin which is using add-multiply as body.
Body of TensorIterator could be any op so I think more is better. But structure of test class, parameters are so different. Just co-existing of both classes is easy but converting my one seems to take time.

**Comment by ilyachur** [comment]
> > > @ilyachur @iefode I met a confliction of tensor_iterator.cpp with #9752. The PR created new op reference test in template plugin(docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp). Then do I need to merge so both test cases exist in the file? Or close this PR without merge?
> > 
> > 
> > Please check coverage of tests and if we need to have more test cases, please extend tests.
> 
> @ilyachur My approach is similar with single layer test of functional shared_test_classes. It also uses recurrence ops. The new one is similar with single layer test of cpu plugin which is using add-multiply as body. Body of TensorIterator could be any op so I think more is better. But structure of test class, parameters are so different. Just co-existing of both classes is easy but converting my one seems to take time.

Ok, can we just extend test cases? If we need we can add new test class with the comment why do we need this class

**Comment by wilson-seok** [comment]
> > > > @ilyachur @iefode I met a confliction of tensor_iterator.cpp with #9752. The PR created new op reference test in template plugin(docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp). Then do I need to merge so both test cases exist in the file? Or close this PR without merge?
> > > 
> > > 
> > > Please check coverage of tests and if we need to have more test cases, please extend tests.
> > 
> > 
> > @ilyachur My approach is similar with single layer test of functional shared_test_classes. It also uses recurrence ops. The new one is similar with single layer test of cpu plugin which is using add-multiply as body. Body of TensorIterator could be any op so I think more is better. But structure of test class, parameters are so different. Just co-existing of both classes is easy but converting my one seems to take time.
> 
> Ok, can we just extend test cases? If we need we can add new test class with the comment why do we need this class

@ilyachur I named my test class as static test while the existing one is using dynamic shape for input parameters. Completed merging them.

**Comment by wilson-seok** [comment]
@ilya @iefode Could you please review this PR? Thanks!

File: docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp [hunk 4]
```diff
@@ -95,16 +101,2338 @@ TEST_P(ReferenceTILayerTest, TensorIteratorWithHardcodedRefs) {
...
+    reference_tests::Tensor B;
+    reference_tests::Tensor Y;
+    reference_tests::Tensor Ho;
+    reference_tests::Tensor Co;
+    std::string testcaseName;
+};
+
+struct TIStaticInputs : public TIStaticFunctionalBase {
+    std::shared_ptr<ov::Model> create_function(const TensorIteratorStaticParams& params) override {
+        std::vector<std::vector<size_t>> inputShapes;
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp]:
```suggestion
        std::vector<ov::Shape> inputShapes;
```

File: docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp [hunk 5]
```diff
@@ -95,16 +101,2338 @@ TEST_P(ReferenceTILayerTest, TensorIteratorWithHardcodedRefs) {
...
+                    // swap params.batchSize and params.seqLength
+                    std::swap(inputShapes[0][0], inputShapes[0][1]);
+                }
+                auto outer_params = ngraph::builder::makeParams(params.iType, {inputShapes[0], inputShapes[1], inputShapes[2]});
+
+                // 1. Create TensorIterator body.
+                inputShapes[0][params.sequenceAxis] = 1; // sliced dimension
+                auto body_params = ngraph::builder::makeParams(params.iType, {inputShapes[0], inputShapes[1], inputShapes[2]});
+                auto squeeze = std::make_shared<ov::opset8::Squeeze>(body_params[0], axis);
+                ngraph::OutputVector out_vector = {squeeze, body_params[1], body_params[2]};
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp]:
```suggestion
                ov::OutputVector out_vector = {squeeze, body_params[1], body_params[2]};
```

**Review by ilyachur — APPROVED** [review]

File: docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp (hunk 4 above)
Comment (wilson-seok) [reply docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp]: Updated. Thanks!

File: docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp (hunk 5 above)
Comment (wilson-seok) [reply docs/template_plugin/tests/functional/op_reference/tensor_iterator.cpp]: Updated. Thanks!

**Review by iefode — APPROVED** [review]

**Pull request merged**